internal/pkg/agent/vault/vault_linux.go

```diff
@@ -59,42 +65,90 @@ func New(path string, opts ...OptionFunc) (v *Vault, err error) {
 		}
 	}
 
-	key, err := getOrCreateSeed(path, options.readonly)
+	r := &Vault{
+		path:       path,
+		retryDelay: options.retryDelay,
+		lock:       flock.New(filepath.Join(path, lockFile)),
+	}
+
+	err = r.tryLock(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = r.unlock(err)
+	}()
+
+	r.key, err = getOrCreateSeed(path, options.readonly)
 	if err != nil {
 		return nil, fmt.Errorf("could not get seed to create new valt: %w", err)
 	}
 
-	return &Vault{
-		path: path,
-		key:  key,
-	}, nil
+	return r, nil
+}
+
+// try to acquire exclusive lock
+func (v *Vault) tryLock(ctx context.Context) error {
+	_, err := v.lock.TryLockContext(ctx, v.retryDelay)
+	if err != nil {
+		return fmt.Errorf("failed to acquire exclusive lock: %v, err: %w", v.lock.Path(), err)
+	}
+	return err
+}
+
+// try to acquire shared lock
+func (v *Vault) tryRLock(ctx context.Context) error {
+	_, err := v.lock.TryRLockContext(ctx, v.retryDelay)
+	if err != nil {
+		return fmt.Errorf("failed to acquire shared lock: %v, err: %w", v.lock.Path(), err)
+	}
+	return err
 }
 
-// Close closes the valut store
+// unlock unlocks the file lock and preserves the original error if there was a error
+func (v *Vault) unlock(err error) error {
+	unerr := v.lock.Unlock()
+	if err != nil {
+		return err
+	}
+	return unerr
+}
```

Comment:
[Blocker]
Both errors are relevant. Better to use `errors.Join` and return both of them.
You can even just pass the errors to `errors.Join` and it'll take care of omitting the `nil` errors and returning `nil` if all errors are `nil`

```suggestion
func (v *Vault) unlock(err error) error {
	unerr := v.lock.Unlock()
	if err != nil {
		return err
	}
	return errors.Join(unerr, err)
}
```